demos/free_surface/explicit_free_surface.py

```diff
@@ -0,0 +1,187 @@
+
+        self.X = SpatialCoordinate(self.mesh)
+
+        self.initialise_temperature()
+
+        self.initialise_approximation()
+        self.rho0 = self.approximation.rho  # This defaults to rho0 = 1 (dimensionless)
+        self.g = self.approximation.g  # This defaults to g = 1 (dimensionless)
+
+        self.initialise_free_surfaces()
+
+        # timestepping
+        self.mu = Constant(1)  # Viscosity (dimensionless)
+        self.tau0 = Constant(2 * self.kk * self.mu / (self.rho0 * self.g))  # Characteristic time scale (dimensionless)
+        log("tau0", self.tau0)
+
+        self.dt = Constant(dt_factor*self.tau0)  # timestep (dimensionless)
+        log("dt (dimensionless)", self.dt)
+
+        self.time = Constant(0.0)
+        self.max_timesteps = round(10*self.tau0/self.dt)  # Simulation runs for 10 characteristic time scales so end state is close to being fully relaxed
+        log("max_timesteps", self.max_timesteps)
+
+        self.setup_bcs()
+
+        self.setup_nullspaces()
+
+        self.setup_solver()
+
+        self.error = 0
+
+        # analytical function
+        self.update_analytical_free_surfaces()
+        # Create output file and select output_frequency:
+        if do_write:
+            # Write output files in VTK format:
+            self.dump_period = 1
+            log("dump_period ", self.dump_period)
+            self.setup_output_file()
+            self.write_file()
+
+    def setup_mesh(self):
+        # Set up geometry:
+        self.D = 3e6  # Depth of domain in m
+        self.L = self.D  # Length of the domain in m
+        self.L0 = self.D  # characteristic length scale for scaling the equations
+        ny = self.nx
+        self.left_id, self.right_id, self.bottom_id, self.top_id = 1, 2, 3, 4  # Boundary IDs
+        self.mesh = RectangleMesh(self.nx, ny, self.L/self.L0, self.D/self.L0)  # Rectangle mesh generated via firedrake
+        self.ds = ds  # Need this so that cylindrical case can overload this later with CombinedSurfaceMeasure
+
+    def setup_function_space(self):
+        self.Z = MixedFunctionSpace([self.V, self.W])  # Mixed function space for velocity and pressure.
+
+    def setup_variables(self):
+        # Function to store the solutions:
+        self.z = Function(self.Z)  # a field over the mixed function space Z.
+        self.stokes_vars = self.z.subfunctions
+        self.eta = Function(self.Z[1], name="eta")  # Define a free surface function
+
+        # Next rename for output:
+        self.stokes_vars[0].rename("Velocity")
+        self.stokes_vars[1].rename("Pressure")
+
+    def initialise_wavenumber(self):
+        lam_dimensional = self.D/2  # wavelength of load in m
+        self.lam = lam_dimensional/self.L0  # dimensionless lambda
+        self.kk = Constant(2 * pi / self.lam)  # wavenumber (dimensionless)
+
+    def initialise_temperature(self):
+        self.T = Function(self.Q, name="Temperature").assign(0)  # Setup a dummy function for temperature
```

Comment:
Taking this function as an example, I guess I did say this method (of setting a class member rather than returning a value) is fine (and it is!), but the alternative of returning a value is a bit easier to type-check and make sure the inheritance contract is fulfilled.

Having said that, I don't think it's that important in this set of examples where you've already verified the results, and they're all being run (and therefore tested). Just leaving the comment for posterity!

Reply:
Cool thanks for explaining that Angus!